File: packages/lib/src/challenge-streak.ts
/**
 * ALFANUMRIK -- Daily Challenge Streak Logic
 *
 * Pure functions for managing daily challenge streaks:
 * - Streak progression (consecutive day tracking)
 * - Mercy day system (grade-dependent, weekly reset)
 * - Milestone detection (badges at 7, 30, 100 days)
 * - Streak visibility rules
 *
 * All functions are pure (no side effects, no DB calls).
 * Configuration constants come from challenge-config.ts.
 * Grades are always strings per P5.
 */

import {
  STREAK_MILESTONES,
  STREAK_VISIBILITY_THRESHOLD,
  getMercyDaysForGrade,
  type StreakMilestone,
} from './challenge-config';

// ---- Types ----

/** Complete streak state for a student. */
export interface StreakState {
  /** Current consecutive challenge streak. */
  currentStreak: number;
  /** All-time best streak. */
  bestStreak: number;
  /** ISO date string of last completed challenge, or null if never. */
  lastChallengeDate: string | null;
  /** Number of mercy days used in the current week. */
  mercyDaysUsedThisWeek: number;
  /** ISO date string of the Monday that starts the current mercy week, or null. */
  mercyWeekStart: string | null;
  /** Array of earned badge IDs. */
  badges: string[];
}

// ---- Internal Helpers ----

/**
 * Format a Date as a local YYYY-MM-DD string (timezone-safe).
 */
function formatLocalDate(d: Date): string {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, '0');
  const day = String(d.getDate()).padStart(2, '0');
  return `${y}-${m}-${day}`;
}

/**
 * Calculate the number of calendar days between two date strings.
 * Returns positive if todayStr is after lastDateStr.
 * Uses UTC parsing to avoid timezone drift.
 */
function dayDifference(lastDateStr: string, todayStr: string): number {
  // Parse as UTC to get consistent calendar-day arithmetic
  const last = Date.UTC(
    parseInt(lastDateStr.slice(0, 4)),
    parseInt(lastDateStr.slice(5, 7)) - 1,
    parseInt(lastDateStr.slice(8, 10))
  );
  const today = Date.UTC(
    parseInt(todayStr.slice(0, 4)),
    parseInt(todayStr.slice(5, 7)) - 1,
    parseInt(todayStr.slice(8, 10))
  );
  return Math.round((today - last) / (1000 * 60 * 60 * 24));
}

/**
 * Get the Monday of the week containing the given date string.
 * Uses ISO week (Monday = start of week).
 * Returns a local YYYY-MM-DD string (timezone-safe).
 */
function getMondayOfWeek(dateStr: string): string {
  const d = new Date(dateStr + 'T12:00:00'); // noon to avoid DST edge cases
  const day = d.getDay(); // 0=Sun, 1=Mon, ...
  const diff = day === 0 ? 6 : day - 1; // days since Monday
  d.setDate(d.getDate() - diff);
  return formatLocalDate(d);
}

// ---- Streak Processing ----

/**
 * Processes a daily challenge completion and returns the updated streak state.
 *
 * Rules:
 * - Same day: no change
 * - Consecutive day (diff=1): increment streak
 * - Missed 1 day (diff=2): use mercy if available, otherwise break streak
 * - Missed 2+ days (diff>=3): break streak (reset to 1)
 * - First ever challenge: set streak to 1
 * - Updates bestStreak if currentStreak exceeds it
 * - Resets mercy counter on new week (Monday-based)
 *
 * @param state Current streak state
 * @param todayStr ISO date string for today (e.g., "2026-04-16")
 * @param grade Student's grade as string ("6" through "12") per P5
 * @returns Updated StreakState
 */
export function processStreakDay(state: StreakState, todayStr: string, grade: string): StreakState {
  const newState: StreakState = { ...state, badges: [...state.badges] };

  // First ever challenge
  if (!state.lastChallengeDate) {
    newState.currentStreak = 1;
    newState.bestStreak = Math.max(state.bestStreak, 1);
    newState.lastChallengeDate = todayStr;
    newState.mercyWeekStart = getMondayOfWeek(todayStr);
    return newState;
  }

  const diff = dayDifference(state.lastChallengeDate, todayStr);

  // Same day -- no change
  if (diff === 0) {
    return newState;
  }

  // Reset mercy counter if we're in a new week
  const currentMonday = getMondayOfWeek(todayStr);
  if (state.mercyWeekStart !== currentMonday) {
    newState.mercyDaysUsedThisWeek = 0;
    newState.mercyWeekStart = currentMonday;
  }

  if (diff === 1) {
    // Consecutive day -- simple increment
    newState.currentStreak = state.currentStreak + 1;
  } else if (diff === 2) {
    // Missed exactly 1 day -- check mercy eligibility
    if (checkMercyEligibility(newState.mercyDaysUsedThisWeek, 1, grade)) {
      newState.currentStreak = state.currentStreak + 1;
      newState.mercyDaysUsedThisWeek += 1;
    } else {
      // Mercy not available -- streak breaks
      newState.currentStreak = 1;
    }
  } else {
    // Missed 2+ days (diff >= 3) -- streak breaks
    newState.currentStreak = 1;
  }

  newState.lastChallengeDate = todayStr;
  newState.bestStreak = Math.max(newState.bestStreak, newState.currentStreak);

  return newState;
}

// ---- Mercy Eligibility ----

/**
 * Checks whether a mercy day can be used.
 *
 * @param mercyUsedThisWeek Number of mercy days already used this week
 * @param daysMissed Number of days missed (must be exactly 1 for mercy to apply)
 * @param grade Student's grade as string per P5
 * @returns true if mercy can be used
 */
export function checkMercyEligibility(
  mercyUsedThisWeek: number,
  daysMissed: number,
  grade: string
): boolean {
  // Mercy only applies to exactly 1 missed day
  if (daysMissed !== 1) return false;

  const allowedMercyDays = getMercyDaysForGrade(grade);
  return mercyUsedThisWeek < allowedMercyDays;
}

// ---- Milestone Detection ----

/**
 * Detects newly crossed streak milestones.
 * Returns milestones that were crossed (previousStreak < threshold <= newStreak)
 * and are not already in the student's badge list.
 *
 * @param previousStreak Streak before the current update
 * @param newStreak Streak after the current update
 * @param existingBadges Array of badge IDs already earned
 * @returns Array of newly earned StreakMilestone objects
 */
export function detectMilestones(
  previousStreak: number,
  newStreak: number,
  existingBadges: string[]
): StreakMilestone[] {
  const existingSet = new Set(existingBadges);

  return STREAK_MILESTONES.filter(milestone => {
    // Milestone must be newly crossed: was below before, now at or above
    const wasBelowBefore = previousStreak < milestone.days;
    const isAtOrAboveNow = newStreak >= milestone.days;
    const notAlreadyEarned = !existingSet.has(milestone.badgeId);

    return wasBelowBefore && isAtOrAboveNow && notAlreadyEarned;
  });
}

// ---- Streak Visibility ----

/**
 * Returns whether the streak should be displayed to the student.
 * Streaks below the threshold are not shown (avoids "1-day streak" clutter).
 *
 * @param streak Current streak count
 * @returns true if the streak should be shown in the UI
 */
export function shouldShowStreak(streak: number): boolean {
  return streak >= STREAK_VISIBILITY_THRESHOLD;
}

File: mobile/lib/core/game/challenge_streak.dart
// Daily Challenge streak logic — pure Dart port of
// `packages/lib/src/challenge-streak.ts`. Mirrors that file's logic
// EXACTLY (streak progression, grade-dependent weekly mercy days,
// milestone detection at 7/30/100 days, visibility threshold).
//
// All functions are pure (no side effects, no DB calls). Configuration
// constants come from `challenge_config.dart`. Grades are always strings
// per P5.
library;

import '../../data/models/challenge_models.dart';
import 'challenge_config.dart';

// ---- Internal Helpers ----

/// Format a [DateTime] as a local YYYY-MM-DD string (timezone-safe).
String _formatLocalDate(DateTime d) {
  final y = d.year.toString().padLeft(4, '0');
  final m = d.month.toString().padLeft(2, '0');
  final day = d.day.toString().padLeft(2, '0');
  return '$y-$m-$day';
}

/// Calculate the number of calendar days between two "YYYY-MM-DD" date
/// strings. Returns positive if [todayStr] is after [lastDateStr]. Uses UTC
/// parsing to avoid timezone drift (mirrors the web's `Date.UTC` approach).
int _dayDifference(String lastDateStr, String todayStr) {
  final last = DateTime.utc(
    int.parse(lastDateStr.substring(0, 4)),
    int.parse(lastDateStr.substring(5, 7)),
    int.parse(lastDateStr.substring(8, 10)),
  );
  final today = DateTime.utc(
    int.parse(todayStr.substring(0, 4)),
    int.parse(todayStr.substring(5, 7)),
    int.parse(todayStr.substring(8, 10)),
  );
  return today.difference(last).inDays;
}

/// Get the Monday of the week containing the given "YYYY-MM-DD" date
/// string. Uses ISO week (Monday = start of week). Returns a local
/// YYYY-MM-DD string (timezone-safe; noon anchor avoids DST edge cases,
/// mirrored from the web implementation even though Dart has no DST bugs
/// in practice for date-only math).
String _getMondayOfWeek(String dateStr) {
  final d = DateTime.parse('${dateStr}T12:00:00');
  final day = d.weekday; // Dart: 1=Mon .. 7=Sun
  final diff = day - 1; // days since Monday
  final monday = d.subtract(Duration(days: diff));
  return _formatLocalDate(monday);
}

// ---- Streak Processing ----

/// Processes a daily challenge completion and returns the updated streak
/// state.
///
/// Rules:
/// - Same day: no change
/// - Consecutive day (diff=1): increment streak
/// - Missed 1 day (diff=2): use mercy if available, otherwise break streak
/// - Missed 2+ days (diff>=3): break streak (reset to 1)
/// - First ever challenge: set streak to 1
/// - Updates bestStreak if currentStreak exceeds it
/// - Resets mercy counter on new week (Monday-based)
///
/// [todayStr] is an ISO date string ("YYYY-MM-DD"). [grade] is a string
/// ("6" through "12") per P5.
StreakState processStreakDay(StreakState state, String todayStr, String grade) {
  // First ever challenge.
  if (state.lastChallengeDate == null) {
    return state.copyWith(
      currentStreak: 1,
      bestStreak: state.bestStreak < 1 ? 1 : state.bestStreak,
      lastChallengeDate: todayStr,
      mercyWeekStart: _getMondayOfWeek(todayStr),
    );
  }

  final diff = _dayDifference(state.lastChallengeDate!, todayStr);

  // Same day — no change.
  if (diff == 0) return state;

  // Reset mercy counter if we're in a new week.
  final currentMonday = _getMondayOfWeek(todayStr);
  var mercyUsed = state.mercyDaysUsedThisWeek;
  var mercyWeekStart = state.mercyWeekStart;
  if (state.mercyWeekStart != currentMonday) {
    mercyUsed = 0;
    mercyWeekStart = currentMonday;
  }

  int newStreak;
  if (diff == 1) {
    // Consecutive day — simple increment.
    newStreak = state.currentStreak + 1;
  } else if (diff == 2) {
    // Missed exactly 1 day — check mercy eligibility.
    if (checkMercyEligibility(mercyUsed, 1, grade)) {
      newStreak = state.currentStreak + 1;
      mercyUsed += 1;
    } else {
      newStreak = 1;
    }
  } else {
    // Missed 2+ days (diff >= 3) — streak breaks.
    newStreak = 1;
  }

  return state.copyWith(
    currentStreak: newStreak,
    bestStreak: newStreak > state.bestStreak ? newStreak : state.bestStreak,
    lastChallengeDate: todayStr,
    mercyDaysUsedThisWeek: mercyUsed,
    mercyWeekStart: mercyWeekStart,
  );
}

// ---- Mercy Eligibility ----

/// Checks whether a mercy day can be used. Mercy only applies to exactly 1
/// missed day. [grade] is a string per P5.
bool checkMercyEligibility(int mercyUsedThisWeek, int daysMissed, String grade) {
  if (daysMissed != 1) return false;
  final allowedMercyDays = getMercyDaysForGrade(grade);
  return mercyUsedThisWeek < allowedMercyDays;
}

// ---- Milestone Detection ----

/// Detects newly crossed streak milestones. Returns milestones that were
/// crossed (previousStreak < threshold <= newStreak) and are not already
/// in the student's badge list.
List<StreakMilestone> detectMilestones(
  int previousStreak,
  int newStreak,
  List<String> existingBadges,
) {
  final existingSet = existingBadges.toSet();

  return kStreakMilestones.where((milestone) {
    final wasBelowBefore = previousStreak < milestone.days;
    final isAtOrAboveNow = newStreak >= milestone.days;
    final notAlreadyEarned = !existingSet.contains(milestone.badgeId);
    return wasBelowBefore && isAtOrAboveNow && notAlreadyEarned;
  }).toList(growable: false);
}

// ---- Streak Visibility ----

/// Returns whether the streak should be displayed to the student. Streaks
/// below the threshold are not shown (avoids "1-day streak" clutter).
bool shouldShowStreak(int streak) => streak >= kStreakVisibilityThreshold;
